### ISSUE TITLE
Tx.checkScripts incorrectly throws error when finalizing tx containing reference scripts (CIP33)

### ISSUE THREAD

**Issue by inimrod** [Hyperion-BT/helios#98]
When calling Tx.finalize() on a tx that contains a CIP33 reference script in one of its outputs, an error gets thrown by the `Tx.checkScripts()` internal method:

`too many scripts included, not all are needed`

Apparently, the reference scripts in the outputs are [not counted](https://github.com/Hyperion-BT/helios/blob/96f802bb5e7ec42188ee60e5a615ec5935309a82/src/tx-builder.js#L760) among the `wantedScripts`. 

The subject TxOutput is constructed this way:

```
const tx = new Tx();
const refOutput = new TxOutput(
    changeAddr,
    refUtxoValue,
    null,
    refScriptCompiled
);
tx.addOutput(refOutput);
```

When [checkScripts](https://github.com/Hyperion-BT/helios/blob/96f802bb5e7ec42188ee60e5a615ec5935309a82/src/tx-builder.js#L1176) is commented out, the transaction gets finalized as expected and successfully submitted on-chain.

**Comment by christianschmitz** [comment]
I'm not sure if this is related to the ref script in the output.

Usually this error means that more scripts are witnessing the transaction than necessary. 

Could you paste the output of `JSON.stringify(tx.dump(), undefined, 4)` below (your version with checkScripts commented out)

**Comment by christianschmitz** [comment]
v0.15.5 hopefully fixes this

**Comment by inimrod** [comment]
> Could you paste the output of `JSON.stringify(tx.dump(), undefined, 4)` below? (your version with checkScripts commented out)

Not sure if you still need it, but here it is:

```
{
    "body": {
        "inputs": [
            {
                "outputId": "3199d...87b08#0",
                "output": {
                    "address": {
                        "hex": "0023c239...a0ee7",
                        "bech32": "addr_test1qq...drpm"
                    },
                    "value": {
                        "lovelace": "1202490",
                        "assets": {
                            "d517c...2d24a": {
                                "52686...57261": "1"
                            }
                        }
                    },
                    "datum": null,
                    "refScript": null
                }
            },
            {
                "outputId": "4104e...857c9#1",
                "output": {
                    "address": {
                        "hex": "0023c239...a0ee7",
                        "bech32": "addr_test1qq...drpm"
                    },
                    "value": {
                        "lovelace": "2500000",
                        "assets": {
                            "a92f8...c6651": {
                                "55547...53038": "1"
                            }
                        }
                    },
                    "datum": null,
                    "refScript": null
                }
            },
            {
                "outputId": "5e223...b1fdb#0",
                "output": {
                    "address": {
                        "hex": "0023c239...a0ee7",
                        "bech32": "addr_test1qq...drpm"
                    },
                    "value": {
                        "lovelace": "10000000000",
                        "assets": {}
                    },
                    "datum": null,
                    "refScript": null
                }
            },
            {
                "outputId": "ac4cc...f2ea0#1",
                "output": {
                    "address": {
                        "hex": "0023c239...a0ee7",
                        "bech32": "addr_test1qq...drpm"
                    },
                    "value": {
                        "lovelace": "1590390",
                        "assets": {
                            "0dafa...884e1": {
                                "424843": "3206364",
                                "485053": "4000000"
                            },
                            "9afaf...2cb9e": {
                                "66...344": "46000000"
                            },
                            "d9610...0ef0b": {
                                "4b726...97073": "1"
                            }
                        }
                    },
                    "datum": null,
                    "refScript": null
                }
            }
        ],
        "outputs": [
            {
                "address": {
                    "hex": "102eb...a0ee7",
                    "bech32": "addr_test1...g8mmn"
                },
                "value": {
                    "lovelace": "2685130",
                    "assets": {
                        "50ac6...f9c33": {
                            "766f7...46f72": "1"
                        }
                    }
                },
                "datum": null,
                "refScript": "5901495901...2c00240021"
            },
            {
                "address": {
                    "hex": "102eb...a0ee7",
                    "bech32": "addr_test1...g8mmn"
                },
                "value": {
                    "lovelace": "11206000",
                    "assets": {
                        "50ac6...f9c33": {
                            "766f7...46f72": "1"
                        }
                    }
                },
                "datum": null,
                "refScript": "5909005908......600120011"
            },
            {
                "address": {
                    "hex": "102eb...a0ee7",
                    "bech32": "addr_test1...g8mmn"
                },
                "value": {
                    "lovelace": "8089870",
                    "assets": {
                        "50ac6...f9c33": {
                            "766f7...96379": "1"
                        }
                    }
                },
                "datum": null,
                "refScript": "5906345906...001200175d"
            },
            {
                "address": {
                    "hex": "10ff2...a0ee7",
                    "bech32": "addr_test1...q36qy"
                },
                "value": {
                    "lovelace": "1607630",
                    "assets": {
                        "50ac6...f9c33": {
                            "766f71...4617465": "1"
                        }
                    }
                },
                "datum": { ... },
                "refScript": null
            },
            {
                "address": {
                    "hex": "0023c239...a0ee7",
                    "bech32": "addr_test1qq...drpm"
                },
                "value": {
                    "lovelace": "1999840",
                    "assets": {
                        "d517c...2d24a": {
                            "52686...57261": "1"
                        },
                        "a92f8...c6651": {
                            "55547...53038": "1"
                        },
                        "0dafa...884e1": {
                            "42...43": "3206364",
                            "48...53": "4000000"
                        },
                        "9afaf...2cb9e": {
                            "66...344": "46000000"
                        },
                        "d9610...0ef0b": {
                            "4b726...97073": "1"
                        }
                    }
                },
                "datum": null,
                "refScript": null
            },
            {
                "address": {
                    "hex": "0023c239...a0ee7",
                    "bech32": "addr_test1qq...drpm"
                },
                "value": {
                    "lovelace": "9979255540",
                    "assets": {}
                },
                "datum": null,
                "refScript": null
            }
        ],
        "fee": "448870",
        "lastValidSlot": null,
        "firstValidSlot": null,
        "minted": {
            "50ac6...f9c33": {
                "766f7...96379": "1",
                "766f7...46f72": "1",
                "766f7...17465": "1",
                "766f7...46f72": "1"
            }
        },
        "metadataHash": null,
        "scriptDataHash": "8055a...6d82b",
        "collateral": [
            {
                "outputId": "5e223...b1fdb#0",
                "output": {
                    "address": {
                        "hex": "0023c239...a0ee7",
                        "bech32": "addr_test1qq...drpm"
                    },
                    "value": {
                        "lovelace": "10000000000",
                        "assets": {}
                    },
                    "datum": null,
                    "refScript": null
                }
            }
        ],
        "signers": [
            "ba887...9be43"
        ],
        "collateralReturn": {
            "address": {
                "hex": "0023c239...a0ee7",
                "bech32": "addr_test1qq...drpm"
            },
            "value": {
                "lovelace": "9999286500",
                "assets": {}
            },
            "datum": null,
            "refScript": null
        },
        "refInputs": []
    },
    "witnesses": {
        "signatures": [
            {
                "pubKeyHash": "23c23...139f7",
                "signature": "ca7d1...83050ec803"
            },
            {
                "pubKeyHash": "ba887...9be43",
                "signature": "3b6ca...8d8a66f807"
            }
        ],
        "datums": [],
        "redeemers": [
            {
                "data": "0{}",
                "exUnits": {
                    "mem": "28137",
                    "cpu": "9330940"
                },
                "type": 1,
                "typeName": "minting",
                "mphIndex": 0
            }
        ],
        "nativeScripts": [],
        "scripts": [
            "59014d5901...800480041"
        ],
        "refScripts": [
            "5901495901...2c00240021",
            "5909005908...1600120011",
            "5906345906...001200175d"
        ]
    },
    "valid": true,
    "metadata": null
}
```

**Comment by christianschmitz** [comment]
@inimrod thanks for the very clean tx.dump() :100:

I see a single minting policy, so the attached script will be related to that, and the three refscripts are indeed taken from the outputs but there is no other redeemer.

I thinks this confirms that the change I made for v0.15.5 is indeed the right fix.

**Comment by inimrod** [comment]
Awesome! Works as expected now. 👍

**Comment by inimrod** [comment]
I have to reopen this issue. It seems it's not totally fixed yet.

I wanted to mint some tokens using a policy script which in the same tx I also want to attach into a utxo -- as a reference script -- so that for succeeding mints, I can just reference to the same policy script.

So when building the tx, I called `Tx::addOutput` to add the output containing the minting policy as a **refScript**.

Then later, I called `Tx::attachScript` to add the same minting policy, but this time, as a **script witness** for the current tx.

This triggered an `unpected` error [here](https://github.com/Hyperion-BT/helios/blob/b81b8c2492d52c78022946a9e97ef2333e25f6bf/src/tx-builder.js#L2563).

And when that `assert()` line is commented out, I get the "[too many scripts](https://github.com/Hyperion-BT/helios/blob/b81b8c2492d52c78022946a9e97ef2333e25f6bf/src/tx-builder.js#L814)" error again:
```
Error: too many scripts included, not all are needed
    at Tx.checkScripts (...node_modules/@hyperionbt/helios/helios.js:45848:10)
    at Tx.finalize (...node_modules/@hyperionbt/helios/helios.js:46261:8)
    ...
```

And of course, if I remove my call to `Tx::attachScript`, Helios doesn't complain but the tx gets rejected by cardano-node because of the missing script witness.

**Comment by inimrod** [comment]
I've tried logging `checkScripts` from [here](https://github.com/Hyperion-BT/helios/blob/aabe13d7be49856b926f17c4ba412e2c29663605/src/tx-builder.js#L800C1-L800C1) and [here](https://github.com/Hyperion-BT/helios/blob/aabe13d7be49856b926f17c4ba412e2c29663605/src/tx-builder.js#L812) and this is what I got:

```
starting checkScripts()
scripts:  4
wantedScripts:  3

after removing ref scripts:
scripts:  4
wantedScripts:  3
```
But I haven't figured out why both `scripts` and `refScripts` seem to be counted as script witnesses for the tx.

txdump of successful tx looks like this (2 `script`s and 2 `refScripts` ):

`````
{
    "body": {
        "inputs": [
            {
                "outputId": "cf5fe...69bf7#0",
                "output": {
                    "address": {
                        "hex": "10ff2...a0ee7",
                        "bech32": "addr_test1...q36qy"
                    },
                    "value": {
                        "lovelace": "1599010",
                        "assets": {
                            "50ac6...f9c33": {
                                "766f71...617465": "1"
                            }
                        }
                    },
                    "datum": {
                        "inlineCbor": "9fd8799f58...51ff0000ff",
                        "inlineSchema": {
                            "list": [
                                {
                                    "constructor": 0,
                                    "fields": [
                                        {
                                            "bytes": "eaaafa7b36...fb7632861e"
                                        },
                                        {
                                            "bytes": "000643b0...564f51"
                                        }
                                    ]
                                },
                                {
                                    "constructor": 0,
                                    "fields": [
                                        {
                                            "bytes": "eaaafa7b36...fb7632861e"
                                        },
                                        {
                                            "bytes": "0014df1...0564f51"
                                        }
                                    ]
                                },
                                {
                                    "int": 0
                                },
                                {
                                    "int": 0
                                }
                            ]
                        }
                    },
                    "refScript": null
                }
            },
            {
                "outputId": "cf5fe...69bf7#1",
                "output": {
                    "address": {
                        "hex": "0023c...a0ee7",
                        "bech32": "addr_test1...9drpm"
                    },
                    "value": {
                        "lovelace": "9977845219",
                        "assets": {}
                    },
                    "datum": null,
                    "refScript": null
                }
            }
        ],
        "outputs": [
            {
                "address": {
                    "hex": "102eb...a0ee7",
                    "bech32": "addr_test1...g8mmn"
                },
                "value": {
                    "lovelace": "8951870",
                    "assets": {
                        "50ac6...f9c33": {
                            "766f7...96379": "1"
                        }
                    }
                },
                "datum": {
                    "inlineCbor": "d87...980",
                    "inlineSchema": {
                        "constructor": 0,
                        "fields": []
                    }
                },
                "refScript": "5906f35906...1600120011"
            },
            {
                "address": {
                    "hex": "10ff2...a0ee7",
                    "bech32": "addr_test1...q36qy"
                },
                "value": {
                    "lovelace": "1633490",
                    "assets": {
                        "50ac6...f9c33": {
                            "766f7...17465": "1"
                        }
                    }
                },
                "datum": {
                    "inlineCbor": "9fd8799f58...10b87200ff",
                    "inlineSchema": {
                        "list": [
                            {
                                "constructor": 0,
                                "fields": [
                                    {
                                        "bytes": "eaaafa7b36...fb7632861e"
                                    },
                                    {
                                        "bytes": "000643b0564f51"
                                    }
                                ]
                            },
                            {
                                "constructor": 0,
                                "fields": [
                                    {
                                        "bytes": "eaaafa7b36...fb7632861e"
                                    },
                                    {
                                        "bytes": "0014df10564f51"
                                    }
                                ]
                            },
                            {
                                "int": 1
                            },
                            {
                                "int": 69000000000
                            }
                        ]
                    }
                },
                "refScript": null
            },
            {
                "address": {
                    "hex": "102eb...a0ee7",
                    "bech32": "addr_test1...g8mmn"
                },
                "value": {
                    "lovelace": "1905020",
                    "assets": {
                        "eaaafa7b36...fb7632861e": {
                            "00064...64f51": "1"
                        }
                    }
                },
                "datum": {
                    "inlineCbor": "d8799fa644...6a706701ff",
                    "inlineSchema": {
                        "constructor": 0,
                        "fields": [
                            {
                                "map": [
                                    {
                                        "k": {
                                            "bytes": "6e616d65"
                                        },
                                        "v": {
                                            "bytes": "564...56e"
                                        }
                                    },
                                    {
                                        "k": {
                                            "bytes": "6465736372697074696f6e"
                                        },
                                        "v": {
                                            "bytes": "54686...c6473"
                                        }
                                    },
                                    {
                                        "k": {
                                            "bytes": "7469636b6572"
                                        },
                                        "v": {
                                            "bytes": "564...f51"
                                        }
                                    },
                                    {
                                        "k": {
                                            "bytes": "75726c"
                                        },
                                        "v": {
                                            "bytes": "68747...e696f"
                                        }
                                    },
                                    {
                                        "k": {
                                            "bytes": "64656369...6d616c73"
                                        },
                                        "v": {
                                            "int": 6
                                        }
                                    },
                                    {
                                        "k": {
                                            "bytes": "6c6f676f"
                                        },
                                        "v": {
                                            "bytes": "6874747073...302e6a7067"
                                        }
                                    }
                                ]
                            },
                            {
                                "int": 1
                            }
                        ]
                    }
                },
                "refScript": null
            },
            {
                "address": {
                    "hex": "00c29...45b29",
                    "bech32": "addr_test1...e9tga"
                },
                "value": {
                    "lovelace": "1185250",
                    "assets": {
                        "eaaafa7b36...fb7632861e": {
                            "0014df10564f51": "69000000000"
                        }
                    }
                },
                "datum": null,
                "refScript": null
            },
            {
                "address": {
                    "hex": "0023c...a0ee7",
                    "bech32": "addr_test1...9drpm"
                },
                "value": {
                    "lovelace": "9965277338",
                    "assets": {}
                },
                "datum": null,
                "refScript": null
            }
        ],
        "fee": "491261",
        "lastValidSlot": null,
        "firstValidSlot": null,
        "minted": {
            "50ac6...f9c33": {
                "766f7...96379": "1"
            },
            "eaaafa7b36...fb7632861e": {
                "000643b0...f51": "1",
                "0014df10...f51": "69000000000"
            }
        },
        "metadataHash": null,
        "scriptDataHash": "917ef...231af",
        "collateral": [
            {
                "outputId": "cf5fe...69bf7#1",
                "output": {
                    "address": {
                        "hex": "0023c...a0ee7",
                        "bech32": "addr_test1...9drpm"
                    },
                    "value": {
                        "lovelace": "9977845219",
                        "assets": {}
                    },
                    "datum": null,
                    "refScript": null
                }
            }
        ],
        "signers": [
            "ba887...9be43"
        ],
        "collateralReturn": {
            "address": {
                "hex": "0023c...a0ee7",
                "bech32": "addr_test1...9drpm"
            },
            "value": {
                "lovelace": "9977068133",
                "assets": {}
            },
            "datum": null,
            "refScript": null
        },
        "refInputs": [
            {
                "outputId": "dab9f...58f07#1",
                "output": {
                    "address": {
                        "hex": "102eb...a0ee7",
                        "bech32": "addr_test1...g8mmn"
                    },
                    "value": {
                        "lovelace": "11206000",
                        "assets": {
                            "50ac6...f9c33": {
                                "766f7...46f72": "1"
                            }
                        }
                    },
                    "datum": null,
                    "refScript": "59090...20011"       
                }
            }
        ]
    },
    "witnesses": {
        "signatures": [
            {
                "pubKeyHash": "23c23...139f7",
                "signature": "1aee1...6b40f"
            },
            {
                "pubKeyHash": "ba887...9be43",
                "signature": "ab933...59a04"
            }
        ],
        "datums": [],
        "redeemers": [
            {
                "data": "0{}",
                "exUnits": {
                    "mem": "28137",
                    "cpu": "9330940"
                },
                "type": 1,
                "typeName": "minting",
                "mphIndex": 0
            },
            {
                "data": "0{}",
                "exUnits": {
                    "mem": "492420",
                    "cpu": "177186561"
                },
                "type": 1,
                "typeName": "minting",
                "mphIndex": 1
            },
            {
                "data": "0{}",
                "exUnits": {
                    "mem": "605552",
                    "cpu": "222881734"
                },
                "type": 0,
                "typeName": "spending",
                "inputIndex": 0
            }
        ],
        "nativeScripts": [],
        "scripts": [
            "59014d5901...5800480041",
            "5906f35906...1600120011"
        ],
        "refScripts": [
            "5909005908...1600120011",
            "5906f35906...1600120011"
        ]
    },
    "valid": true,
    "metadata": null
}
`````

**Comment by inimrod** [comment]
@christianschmitz [this](https://github.com/Hyperion-BT/helios/blob/aabe13d7be49856b926f17c4ba412e2c29663605/src/tx-builder.js#L634) might be the root of the issue:

```
/**
 * Add a `TxOutput` instance to the transaction being built.
 * 
 * Mutates the transaction.
 * Only available during transaction building.
 * Returns the transaction instance so build methods can be chained.
 * @param {TxOutput} output 
 * @returns {Tx}
 */
addOutput(output) {
	assert(!this.#valid);
	
	// min lovelace is checked during build, because 
	this.#body.addOutput(output);

	// a refScript can potentially be used for evaluation alsready
	if (output.refScript) {
		this.#witnesses.attachRefScript(output.refScript);
	}

	return this;
}
```

Not quite sure about this, but as far as I can tell, `refScripts` attached to outputs are not yet needed as witnesses to the current tx. So they shouldn't be added to the current tx witnesses. And if they are really intended to be used in the tx also (like what I'm trying to do above), they should also be added via `Tx::attachScript`.

**Comment by christianschmitz** [comment]
> If I remove my call to `Tx::attachScript`, Helios doesn't complain but of course the tx gets rejected by cardano-node because of the missing script witness.

That's interesting, that means that automatically attaching the script in the ouput as a refScript is indeed pointless.

I'll remove that line. Sadly I don't remember why I put it there in the first place, probably a quick fix for someone. Hopefully they find out that whatever they are doing now will be broken and then I can provide a proper fix.

**Comment by inimrod** [comment]
confirmed fixed in 0.15.10
